core/src/Kokkos_Graph.hpp

```diff
@@ -0,0 +1,288 @@
+  // <editor-fold desc="public member types"> {{{2
+
+  using execution_space = ExecutionSpace;
+  using graph           = Graph<ExecutionSpace>;
+  using graph_builder   = GraphBuilder;
+
+  // </editor-fold> end public member types }}}2
+  //----------------------------------------------------------------------------
+
+ private:
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="friends"> {{{2
+
+  friend struct Kokkos::Impl::GraphAccess;
+
+  // </editor-fold> end friends }}}2
+  //----------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="private data members"> {{{2
+
+  using graph_impl_t    = Kokkos::Impl::GraphImpl<ExecutionSpace>;
+  using root_node_ref_t = typename graph_impl_t::root_node_impl_t::node_ref_t;
+  root_node_ref_t m_root;
+
+  // </editor-fold> end private data members }}}2
+  //----------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="private ctors"> {{{2
+
+  // Note: only create_graph() uses this constructor, but we can't just make
+  // that a friend instead of GraphAccess because of the way that friend
+  // function template injection works.
+  explicit GraphBuilder(root_node_ref_t arg_root)
+      : m_root(std::move(arg_root)) {}
+
+  // </editor-fold> end private ctors }}}2
+  //----------------------------------------------------------------------------
+
+ public:
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="ctors, destructor, and assignment"> {{{2
+
+  // Rule of 6 for copy constructible
+
+  GraphBuilder() noexcept               = default;
+  GraphBuilder(GraphBuilder const&)     = default;
+  GraphBuilder(GraphBuilder&&) noexcept = default;
+  GraphBuilder& operator=(GraphBuilder const&) = default;
+  GraphBuilder& operator=(GraphBuilder&&) noexcept = default;
+
+  ~GraphBuilder() = default;
+
+  // </editor-fold> end ctors, destructor, and assignment }}}2
+  //----------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="public accessors"> {{{2
+
+  constexpr auto const& get_root() const { return m_root; }
+
+  // </editor-fold> end public accessors }}}2
+  //----------------------------------------------------------------------------
+
+  template <class... PredecessorRefs>
+  // constraints (not intended for subsumption, though...)
+  //   ((remove_cvref_t<PredecessorRefs> is a specialization of
+  //        GraphNodeRef with get_root().get_graph_impl() as its GraphImpl)
+  //      && ...)
+  auto when_all(PredecessorRefs&&... arg_pred_refs) const {
```

Comment:
So this is gonna be a free function since we eliminate the GraphBuilder and just pass in the root node.

Reply:
Addressed in f089cae